_posts/2022-05-02-the-na-bug/the-na-bug.Rmd

```diff
@@ -0,0 +1,111 @@
+
+- The data can be different due to **external factors**. For example, all of our models were trained on data prior to COVID-19, but shortly after the beginning of our silent deployment phase, we began to observe cases of COVID-19 in the hospital.
+
+- The data can be different due to **data entry errors**. For example, a body temperature could incorrectly be entered as 3700 °C instead of 37.00 °C.
+
+- The data can be different due to **selection bias**. For example, during training we excluded patients with really short and really long visits, as they were rare. However, we may encounter these kinds of visits in the live data.
+
+
+## Monitoring labs
+
+We had set up a monitoring dashboard to measure model inputs and model outputs. On close inspection, we made a discovery that was unquestioNAbly odd... no sodium labs had been measured since we had moved to silent testing!
+
+```{r missingna, fig.cap="Daily counts of lab measurements: this includes counts for calcium (CA), chloride (CL), glucose (GLPOC), potassium (K), and sodium (NA)."}
+daily_counts_pre_fix %>%
+    select(-date, -cohort) %>%
+    xts(order.by = daily_counts_pre_fix$date) %>%
+    dygraph(main = "Electrolyte counts")  %>%
+    dyOptions(stackedGraph = TRUE) %>%
+    dyLegend(width = 400)
+```
+
+This made NA sense at all. Sodium is measured in routinely ordered blood tests. It'll usually get ordered alongside other tests (such as calcium, chloride, glucose, and potassium) as part of a [basic metabolic panel](https://www.healthline.com/health/sodium-blood). In Figure \@ref(fig:missingna), we look at the daily counts of labs on units in which CHARTwatch was silently deployed. 
+The other labs were regularly measured, but our pipeline had not detected a single sodium lab. There was NA way sodium would be missing!
+
+## The NA bug
+
+After hours of detective work, we found the issue:
+
+- In R, the programming language we used to develop CHARTwatch, the symbol `NA` stands for "not available" and is used to represent missing data.
+
+- In chemistry, `Na` is the symbol used to represent the chemical element of sodium.
+
+```{r fixedna, fig.cap="Daily counts of lab measurements after fixing the NA bug"}
+data %>%
+    select(-date, -cohort) %>%
+    xts(order.by = data$date) %>%
+    dygraph(main = "Electrolyte counts")  %>%
+    dyOptions(stackedGraph = TRUE) %>%
+    dyLegend(width = 400)
+```
+
+
+Depending on the context, the symbol meant something different! Our data extraction pipeline was interpreting the chemical element `Na` as "not available"!
+
+We were able to fix the _NA bug_ and after re-deploying, sodium counts were back to normal (as seen in Figure \@ref(fig:fixedna)).
+
+The fix was quite simple, but the problem we uncovered lead to plenty of follow-up questions! 
```

Comment:
How did you fix the issue?

Reply:
had to change an argument to one of the functions! I'll include the details in the blog post